### PR TITLE
[JUJU-2350] Drain secrets into current secrets backend

### PR DESCRIPTION
**Pull request by ycliuhw**
This PR introduces the secret drain feature. 
After the model's secret backend changed, Juju now migrates all the applications or units owned secrets to the new backend.
TODO: 

- implement a similar feature for user-created secrets - https://bugs.launchpad.net/juju/+bug/2015967;
- add integration test;

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
- [ ] [Integration tests](https://github.com/juju/juju/tree/develop/tests), with comments saying what you're testing
- [x] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages

## QA steps

- k8s model

```sh
juju model-config secret-backend
auto

juju exec --unit snappass-test/0 'secret-add ownedby=snappass-test/0 --owner unit'
secret:cgr4jseffbas7anrvodg

juju exec --unit snappass-test/0 'secret-add ownedby=snappass-test'
secret:cgr4ju6ffbas7anrvoe0

juju show-secret --reveal cgr4jseffbas7anrvodg
cgr4jseffbas7anrvodg:
  revision: 1
  owner: snappass-test/0
  created: 2023-04-12T06:04:33Z
  updated: 2023-04-12T06:04:33Z
  content:
    owned by: snappass-test/0

juju show-secret --reveal cgr4ju6ffbas7anrvoe0
cgr4ju6ffbas7anrvoe0:
  revision: 1
  owner: snappass-test
  created: 2023-04-12T06:04:43Z
  updated: 2023-04-12T06:04:43Z
  content:
    ownedby: snappass-test

juju exec --unit mariadb-k8s/0 'secret-add ownedby=mariadb-k8s'
secret:cgr4l26ffbas7anrvof0

juju show-secret --reveal cgr4l26ffbas7anrvof0
cgr4l26ffbas7anrvof0:
  revision: 1
  owner: mariadb-k8s
  created: 2023-04-12T06:07:06Z
  updated: 2023-04-12T06:07:06Z
  content:
    ownedby: mariadb-k8s

mkubectl -nt1 get secrets -l 'app.juju.is/created-by in (mariadb-k8s, snappass-test)'
NAME                     TYPE     DATA   AGE
cgr4jseffbas7anrvodg-1   Opaque   1      9m54s
cgr4ju6ffbas7anrvoe0-1   Opaque   1      9m45s
cgr4l26ffbas7anrvof0-1   Opaque   1      7m21s

juju model-config secret-backend=myvault

mkubectl -nt1 get secrets -l 'app.juju.is/created-by in (mariadb-k8s, snappass-test)'
No resources found in t1 namespace.

vault kv list -format json t1-57add6
[
  "cgr4jseffbas7anrvodg-1",
  "cgr4ju6ffbas7anrvoe0-1",
  "cgr4l26ffbas7anrvof0-1"
]

juju model-config secret-backend=auto

vault kv list -format json t1-57add6
{}

mkubectl -nt1 get secrets -l 'app.juju.is/created-by in (mariadb-k8s, snappass-test)'

NAME                     TYPE     DATA   AGE
cgr4l26ffbas7anrvof0-1   Opaque   1      45s
cgr4ju6ffbas7anrvoe0-1   Opaque   1      45s
cgr4jseffbas7anrvodg-1   Opaque   1      39s

```

- IaaS model

```sh
juju model-config secret-backend
auto

juju exec --unit ubuntu/0 'secret-add ownedby=ubuntu/0 --owner unit'
secret:cgr54f2qauo2agu8lfq0

juju show-secret --reveal cgr54f2qauo2agu8lfq0
cgr54f2qauo2agu8lfq0:
  revision: 1
  owner: ubuntu/0
  created: 2023-04-12T06:39:56Z
  updated: 2023-04-12T06:39:56Z
  content:
    ownedby: ubuntu/0

juju exec --unit ubuntu/0 'secret-add ownedby=ubuntu'
secret:cgr55h2qauo2agu8lfqg

juju show-secret --reveal cgr55h2qauo2agu8lfqg
cgr55h2qauo2agu8lfqg:
  revision: 1
  owner: ubuntu
  created: 2023-04-12T06:42:13Z
  updated: 2023-04-12T06:42:13Z
  content:
    ownedby: ubuntu

juju:PRIMARY> db.secretRevisions.find({},{data:1, 'owner-tag': 1, 'value-reference': 1}).pretty()
{
	"_id" : "0f94f920-0372-43bc-8f61-bc119d170c5a:cgr54f2qauo2agu8lfq0/1",
	"data" : {
		"ownedby" : "dWJ1bnR1LzA="
	},
	"owner-tag" : "unit-ubuntu-0",
	"value-reference" : null
}
{
	"_id" : "0f94f920-0372-43bc-8f61-bc119d170c5a:cgr55h2qauo2agu8lfqg/1",
	"data" : {
		"ownedby" : "dWJ1bnR1"
	},
	"owner-tag" : "application-ubuntu",
	"value-reference" : null
}

vault kv list -format json t1-170c5a
{}

juju model-config secret-backend=myvault

juju:PRIMARY> db.secretRevisions.find({},{data:1, 'owner-tag': 1, 'value-reference': 1}).pretty()
{
	"_id" : "0f94f920-0372-43bc-8f61-bc119d170c5a:cgr54f2qauo2agu8lfq0/1",
	"data" : {

	},
	"owner-tag" : "unit-ubuntu-0",
	"value-reference" : {
		"backend-id" : "6436533350f6f810ecba5d6b",
		"revision-id" : "cgr54f2qauo2agu8lfq0-1"
	}
}
{
	"_id" : "0f94f920-0372-43bc-8f61-bc119d170c5a:cgr55h2qauo2agu8lfqg/1",
	"data" : {

	},
	"owner-tag" : "application-ubuntu",
	"value-reference" : {
		"backend-id" : "6436533350f6f810ecba5d6b",
		"revision-id" : "cgr55h2qauo2agu8lfqg-1"
	}
}

vault kv list -format json t1-170c5a
[
  "cgr54f2qauo2agu8lfq0-1",
  "cgr55h2qauo2agu8lfqg-1"
]

juju model-config secret-backend=auto

vault kv list -format json t1-170c5a
{}

juju:PRIMARY> db.secretRevisions.find({},{data:1, 'owner-tag': 1, 'value-reference': 1}).pretty()
{
	"_id" : "0f94f920-0372-43bc-8f61-bc119d170c5a:cgr54f2qauo2agu8lfq0/1",
	"data" : {
		"ownedby" : "dWJ1bnR1LzA="
	},
	"owner-tag" : "unit-ubuntu-0",
	"value-reference" : null
}
{
	"_id" : "0f94f920-0372-43bc-8f61-bc119d170c5a:cgr55h2qauo2agu8lfqg/1",
	"data" : {
		"ownedby" : "dWJ1bnR1"
	},
	"owner-tag" : "application-ubuntu",
	"value-reference" : null
}

```

## Documentation changes

No

## Bug reference

No

### CODE REVIEW

File: apiserver/facades/agent/secretsmanager/secrets.go [hunk 1]
```diff
@@ -321,8 +323,16 @@ func (s *SecretsManagerAPI) getSecretConsumerInfo(consumerTag names.Tag, uriStr
...
+}
+
+func (s *SecretsManagerAPI) getSecretMetadata(
+	filter func(*coresecrets.SecretMetadata, *coresecrets.SecretRevisionMetadata) bool,
+) (params.ListSecretResults, error) {
 	var result params.ListSecretResults
-	filter := state.SecretsFilter{
+	listFilter := state.SecretsFilter{
+		// TODO: there is a bug that operator agents can't get any unit owned secrets!
+		// Because the authTag here is the application tag, but not unit tag.
```
Comment (ycliuhw) [inline apiserver/facades/agent/secretsmanager/secrets.go:334]: created a bug for it: https://bugs.launchpad.net/juju/+bug/2015963

File: worker/secretmigrationworker/worker.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+}
+
+// Worker migrates secrets to the new backend when the model's secret backend has changed.
+type Worker struct {
+	catacomb catacomb.Catacomb
+	config   Config
+}
+
+// TODO: we should do backend connection validation (ping the backend) when we change the secret backend in model config!!!
+// juju model-config secret-backend=myothersecrets
```
Comment (ycliuhw) [inline worker/secretmigrationworker/worker.go:86]: https://bugs.launchpad.net/juju/+bug/2015965

File: worker/secretmigrationworker/worker.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+// Or we should have a secret backend validator worker keeps ping the backend to make sure it's alive.
+// The worker should make the backend as inactive if it's not alive.
+
+// TODO: user created secrets should be migrated on the controller becasue they donot have an owner unit!!
```
Comment (ycliuhw) [inline worker/secretmigrationworker/worker.go]: https://bugs.launchpad.net/juju/+bug/2015967

File: worker/secretmigrationworker/doc.go [hunk 4]
```diff
@@ -0,0 +1,7 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+// Package secretmigrationworker provides a worker running on the agent for
+// migrating secrets to the new active backend
+// when the model changed secret backend.
+package secretmigrationworker
```
Comment (wallyworld) [inline worker/secretmigrationworker/doc.go:7]: We should use "drain", eg "secretdrainworker" etc

File: worker/secretmigrationworker/manifold.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	NewWorker         func(Config) (worker.Worker, error)
+	NewBackendsClient func(jujusecrets.JujuAPIClient) (jujusecrets.BackendsClient, error)
+}
+
+// NewClient returns a new secretmigrationworker facade.
+func NewClient(caller base.APICaller) Facade {
+	return secretsmanager.NewClient(caller)
+}
+
+func NewBackendsClient(facade jujusecrets.JujuAPIClient) (jujusecrets.BackendsClient, error) {
```
Comment (wallyworld) [inline worker/secretmigrationworker/manifold.go:31]: comment

File: worker/secretmigrationworker/worker.go [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+// Logger represents the methods used by the worker to log information.
+type Logger interface {
+	Tracef(string, ...interface{})
+	Debugf(string, ...interface{})
+	Warningf(string, ...interface{})
+	Infof(string, ...interface{})
+	Errorf(string, ...interface{})
+	Criticalf(string, ...interface{})
+}
```
Comment (wallyworld) [inline worker/secretmigrationworker/worker.go:32]: Do we need all these

File: worker/secretmigrationworker/worker.go [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+
+// Facade instances provide a set of API for the worker to deal with secret migration task changes.
+type Facade interface {
+	WatchSecretBackendChanged() (watcher.NotifyWatcher, error)
+	GetSecretsToMigrate() ([]coresecrets.SecretMetadataForMigration, error)
```
Comment (wallyworld) [inline worker/secretmigrationworker/worker.go:37]: GetSecretsToDrain

File: worker/secretmigrationworker/worker.go [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+	Debugf(string, ...interface{})
+	Warningf(string, ...interface{})
+	Infof(string, ...interface{})
+	Errorf(string, ...interface{})
+	Criticalf(string, ...interface{})
+}
+
+// Facade instances provide a set of API for the worker to deal with secret migration task changes.
+type Facade interface {
+	WatchSecretBackendChanged() (watcher.NotifyWatcher, error)
```
Comment (wallyworld) [inline worker/secretmigrationworker/worker.go:36]: The 3 methods here should be on a separate facade. This worker should have 2 facades  - a secrets drain facade with the 3 new methods on it and a JujuAPIClient for processing the changes.

**Review by wallyworld — COMMENTED** [review]
Looks great, main issue is the facade design - there should be a new facade with the methods specific to the drain requirements

File: worker/secretmigrationworker/worker.go (hunk 2 above)
Comment (ycliuhw) [reply worker/secretmigrationworker/worker.go]: Done.

File: worker/secretdrainworker/worker.go [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if len(secrets) == 0 {
+				continue
+			}
+			backends, err := w.config.SecretsBackendGetter()
+			if err != nil {
+				return errors.Trace(err)
+			}
```
Comment (wallyworld) [inline worker/secretdrainworker/worker.go:121]: Can we do this up above outside the for loop?

**Review by wallyworld — APPROVED** [review]
Thank you

File: worker/secretdrainworker/worker.go (hunk 9 above)
Comment (ycliuhw) [reply worker/secretdrainworker/worker.go:121]: We may have added new backends after the worker started.

**Pull request merged**